ByRef closures

ncave opened the pull request:
@alexswan10k
> closures in our implementation are capture by move, not by reference

But do they have to be, though? This is passing all tests, including the mutable one.

alexswan10k commented:
Sure. Here is one:

```
let returnClosure () =
    let a = { Value = "a" }
    let b = { Value = "b" }
    fun x -> a.Value + b.Value + x

[<Fact>]
let ``Closure actually owns internals`` () =
    let cl = returnClosure()
    cl "x" |> equal "abx"
```

Now, this doesn't even compile lol, but the return signature should be the closure syntax `impl Fn(&Rc<str>) -> Rc<str>` and not the function pointer syntax `fn(a: &Rc<str>) -> Rc<str>` (the isParamType switch thing has obviously not yet been generalized for return types also). If you do this, it will compile. Notice how when you remove "move", everything breaks? This is because a and b are no longer owned by the closure, so they are destroyed as soon as returnClosure returns.

ncave commented:
@alexswan10k Makes sense, I switched back to capturing by value, and made mutable let bindings wrapped.

What works now in this PR:
- Capturing mutable variables works.
- Void return type closures work.
- Proper closure return types (`Rc<dyn Fn>`).
- Multiple (conditional) closure returns.

I added a few commented tests for things that don't work yet.
(These are not blokers for this PR as they never did work, we just didn't have the tests to show it before):
- Closures that capture value-type function parameters have lifetime issues (see `ClosureTests`).
  We may have to clone those value-type ref idents too, before the closure captures them by value.
- Mutable reference-type record fields create issues with deriving `Clone`, `PartialEq`, etc. (see RecordTests).
  We might need a custom Cell type that properly implements `Clone`, `PartialEq`, etc.
  Or something else, but just `Rc`-wrapping doesn't help much.
  
Anyway, if you have a bit of time, please review and let me know if you're ok with merging this.

alexswan10k commented:
By the way, on a side note can the impl trait syntax be used everywhere? I thought it was quite fussy and could only be used in params and function signatures.

ncave commented:
@alexswan10k 
> can the impl trait syntax be used everywhere? I thought it was quite fussy and could only be used in params and function signatures.

Yes, impl syntax is fussy and can only be used in input params, not even in return types if you need potentially different types coming out with the same signature (e.g. different closures from branches, see `ClosureTests`). That's why return types are now boxed (`Rc<dyn Fn()>`).

ncave commented:
@alexswan10k To illustrate the issue, here is it with generic types, e.g.:
```rs
pub fn id<T: Clone>(i: &T) -> Rc<dyn FnOnce() -> T> {
    Rc::from({
        // let i = i.clone(); // <- the right thing to do, but doesn't fix it
        move || (*i) // Error: i has an anonimous lifetime, but needs static
    })
}
```
